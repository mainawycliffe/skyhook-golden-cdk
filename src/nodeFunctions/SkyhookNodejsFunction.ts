import {
  NodejsFunction,
  NodejsFunctionProps,
} from "aws-cdk-lib/aws-lambda-nodejs";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";

/**
 * Skyhook Nodejs Function
 *
 * Adds some defaults to the standard
 * [NodejsFunction](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda-nodejs.NodejsFunction) that we
 * prefer to use, including higher memory and minifying the code.
 *
 * Note that we highly recommend having esbuild in your project dev dependencies as cdk will then compile your function
 * using esbuild, which is much faster than webpack.
 */
export default class SkyhookNodejsFunction extends NodejsFunction {
  constructor(scope: Construct, id: string, props: NodejsFunctionProps) {
    // removing out bundling from other props to allow minify to be true unless explicitly set to false
    const { bundling, ...propsWithoutBundling } = props;
    super(scope, id, {
      // Skyhook defaults
      bundling: { minify: true, ...bundling },
      memorySize: 1024, // We have found this to typically represent a good balance of performance vs price.
      runtime: Runtime.NODEJS_14_X, // Currently CDK defaults to 12
      // Given props
      ...propsWithoutBundling,
    });
  }
}
