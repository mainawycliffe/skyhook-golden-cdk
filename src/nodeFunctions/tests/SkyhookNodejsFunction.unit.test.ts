import { Stack } from 'aws-cdk-lib/core';
import { SynthUtils } from '@aws-cdk/assert';
import SkyhookNodejsFunction from '../SkyhookNodejsFunction';
import '@aws-cdk/assert/jest';

const testStack = new Stack();
new SkyhookNodejsFunction(testStack, 'testFunction', { entry: 'dummyEntry' });

it('creates a Lambda Function with the expected defaults', () => {
  // Options at https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html
  expect(testStack).toHaveResourceLike('AWS::Lambda::Function', {
    MemorySize: 1024,
    Runtime: 'nodejs20.x',
  });
});

it('creates resources matching the snapshot', () => {
  const cfn = SynthUtils.toCloudFormation(testStack);
  expect(cfn).toMatchSnapshot();
});
