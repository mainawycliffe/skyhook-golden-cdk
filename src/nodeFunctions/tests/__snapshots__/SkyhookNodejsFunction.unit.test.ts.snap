// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
{
  "Resources": {
    "testFunction483F4CBE": {
      "DependsOn": [
        "testFunctionServiceRoleFEC29B6F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-lambda-nodejs-mock-fake-bucket",
          "S3Key": "cdk-lambda-nodejs-mock-fake-lambda.ts",
        },
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "testFunctionServiceRoleFEC29B6F",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testFunctionServiceRoleFEC29B6F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
